codegen-server/src/main/kotlin/software/amazon/smithy/rust/codegen/server/smithy/protocols/ServerHttpBoundProtocolGenerator.kt

```diff
@@ -551,25 +561,47 @@ private class ServerHttpBoundProtocolTraitImplGenerator(
         }
     }
 
+    private fun serverRenderHttpResponseCode(
+        defaultCode: Int
+    ): Writable {
+        return writable {
+            rustTemplate(
+                """
+                let status = $defaultCode;
+                let http_status: u16 = status.try_into()
+                    .map_err(|_| #{ResponseRejection}::InvalidHttpStatusCode)?;
+                builder = builder.status(http_status);
+                """.trimIndent(),
+                *codegenScope,
+            )
+        }
+    }
+
     private fun serverRenderResponseCodeBinding(
         binding: HttpBindingDescriptor
     ): Writable {
         check(binding.location == HttpLocation.RESPONSE_CODE)
+
         return writable {
             val memberName = symbolProvider.toMemberName(binding.member)
-            // TODO(https://github.com/awslabs/smithy-rs/issues/1229): This code is problematic for two reasons:
-            // 1. We're not falling back to the `http` trait if no `output.$memberName` is `None`.
-            // 2. It only works when `output.$memberName` is of type `Option<i32>`.
+            rust("let status = output.$memberName")
+            if (symbolProvider.toSymbol(binding.member).isOptional()) {
+                rustTemplate(
+                    """
+                    .ok_or(#{ResponseRejection}::MissingHttpStatusCode)?
+                    """.trimIndent(),
+                    *codegenScope,
+                )
+            }
             rustTemplate(
                 """
-                let status = output.$memberName
-                    .ok_or(#{ResponseRejection}::MissingHttpStatusCode)?;
+                ;
```

Comment:
sometimes cleaner ways to generate code that handles this, usually by wrapping in a match instead of chaining with OK or, but this is fine

Reply:
`status` might not be an Option. I'll keep your suggestion in mind, thanks